service/storage_proxy.cc

```diff
@@ -4984,9 +4986,11 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
 
                 // We generate a retry if at least one node reply with count live columns but after merge we have less
                 // than the total number of column we are interested in (which may be < count on a retry).
-                // So in particular, if no host returned count live columns, we know it's not a short read.
-                bool can_send_short_read = rr_opt && rr_opt->is_short_read() && rr_opt->row_count() > 0;
-                if (rr_opt && (can_send_short_read || data_resolver->all_reached_end() || rr_opt->row_count() >= original_row_limit()
+                // So in particular, if no host returned count live columns, we know it's not a short read due to
+                // row or partition limits being exhausted and retry is not needed.
```

Comment:
I don't fully understand the comment, and if I did, I probably wouldn't like it. "short read due to no data" vs "end of data, don't read more pages" should be explicit, not inferred from a complicated set of conditions.

Reply:
Ok, I think it actually is explicit: all_reached_end().